crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -636,21 +767,32 @@ impl<'db> TypeInferenceBuilder<'db> {
 
     fn infer_parameter_with_default_definition(
         &mut self,
-        _parameter_with_default: &ast::ParameterWithDefault,
+        parameter_with_default: &ast::ParameterWithDefault,
         definition: Definition<'db>,
     ) {
         // TODO(dhruvmanila): Infer types from annotation or default expression
-        self.types.definitions.insert(definition, Type::Unknown);
+        // TODO check that default is assignable to parameter type
+        self.infer_parameter_definition(&parameter_with_default.parameter, definition);
     }
 
     fn infer_parameter_definition(
         &mut self,
-        _parameter: &ast::Parameter,
+        parameter: &ast::Parameter,
         definition: Definition<'db>,
     ) {
         // TODO(dhruvmanila): Annotation expression is resolved at the enclosing scope, infer the
         // parameter type from there
-        self.types.definitions.insert(definition, Type::Unknown);
+        let annotated_ty = Type::Unknown;
+        if parameter.annotation.is_some() {
+            self.add_declaration_with_binding(
+                parameter.into(),
+                definition,
+                annotated_ty,
+                annotated_ty,
```

Comment:
Should both of these be `annotated_ty` or should one be `Type::Unknown` because it is the inferred default value type?

Reply:
The default value type doesn't have any influence on either the declared or the inferred type for a parameter. We need to validate that the inferred type of the default value is assignable to the declared type, if any, but that's all that we can do. Consider this function:

```
def f(x: int = 1): ...
```

The declared type for `x` is `int`. The inferred type for the default parameter is `Literal[1]`. But that does not mean that the initial inferred type for `x` in the function's internal scope can be `Literal[1]`! The default value isn't the only possible value; callers are allowed to pass any value of type `int` for `x`. So both the declared and the inferred type for `x` is `int`. The only thing we do with the default value is verify that `Literal[1]` is assignable to `int`, and emit a diagnostic if not.